File: typescript/src/crypto.ts
/**
 * Post-Quantum Cryptography Module for ZAP
 *
 * One library, one implementation: luxcrypto (libluxcrypto FFI).
 * Same PQ crypto from Lux blockchain to AI agents.
 *
 * Backend chain:
 *   1. luxcrypto (libluxcrypto.{dylib,so,dll}) — preferred, FIPS 203/204 via circl
 *   2. Unavailable → throws CryptoError
 *
 * X25519 always uses Web Crypto API (real Curve25519).
 *
 * @example
 * ```typescript
 * import { PQKeyExchange, PQSignature, HybridHandshake } from '@hanzo/zap/crypto';
 *
 * // Key exchange
 * const alice = await PQKeyExchange.generate();
 * const bob = await PQKeyExchange.generate();
 * const [ciphertext, sharedAlice] = await alice.encapsulate(bob.publicKey);
 * const sharedBob = await bob.decapsulate(ciphertext);
 *
 * // Signatures
 * const signer = await PQSignature.generate();
 * const sig = await signer.sign(new TextEncoder().encode('message'));
 * await signer.verify(new TextEncoder().encode('message'), sig);
 *
 * // Hybrid handshake
 * const initiator = await HybridHandshake.initiate();
 * const [responder, response] = await HybridHandshake.respond(initiator.publicData);
 * const sharedInit = await initiator.finalize(response);
 * ```
 */

// ── Backend detection ────────────────────────────────────────────────

let _luxcrypto: typeof import('luxcrypto') | null = null;
let _pqAvailable = false;

try {
  // Dynamic import — only succeeds in Node.js with luxcrypto installed
  _luxcrypto = await import('luxcrypto');
  _pqAvailable = _luxcrypto.LUX_CRYPTO_AVAILABLE;
} catch {
  // luxcrypto not available (browser, missing lib, etc.)
}

// ── Constants ────────────────────────────────────────────────────────

export const MLKEM_PUBLIC_KEY_SIZE = 1184;
export const MLKEM_CIPHERTEXT_SIZE = 1088;
export const MLKEM_SHARED_SECRET_SIZE = 32;
export const MLDSA_PUBLIC_KEY_SIZE = 1952;
export const MLDSA_SIGNATURE_SIZE = 3309;
export const X25519_PUBLIC_KEY_SIZE = 32;
export const HYBRID_SHARED_SECRET_SIZE = 32;

export const PQ_BACKEND: string = _pqAvailable ? 'luxcrypto' : 'unavailable';

/**
 * Cryptographic operation error.
 */
export class CryptoError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CryptoError';
  }
}

/**
 * Check if Web Crypto API is available.
 */
export function isWebCryptoAvailable(): boolean {
  return typeof globalThis.crypto !== 'undefined' &&
         typeof globalThis.crypto.subtle !== 'undefined';
}

/**
 * Check if PQ crypto is available (luxcrypto with libluxcrypto).
 */
export function isPQAvailable(): boolean {
  return _pqAvailable;
}

// ── Interfaces ───────────────────────────────────────────────────────

export interface HybridInitiatorData {
  x25519PublicKey: Uint8Array;
  mlkemPublicKey: Uint8Array;
}

export interface HybridResponderData {
  x25519PublicKey: Uint8Array;
  mlkemCiphertext: Uint8Array;
}

// ── ML-KEM-768 (FIPS 203) ───────────────────────────────────────────

/**
 * ML-KEM-768 Key Encapsulation Mechanism.
 *
 * Implements NIST FIPS 203 ML-KEM-768 via libluxcrypto (cloudflare/circl).
 * Security level: NIST Level 3 (~AES-192 equivalent).
 */
export class PQKeyExchange {
  private readonly _publicKey: Uint8Array;
  private readonly _secretKey: Uint8Array | null;

  private constructor(publicKey: Uint8Array, secretKey: Uint8Array | null) {
    this._publicKey = publicKey;
    this._secretKey = secretKey;
  }

  static async generate(): Promise<PQKeyExchange> {
    if (!_pqAvailable || !_luxcrypto) {
      throw new CryptoError(
        'PQ crypto not available — install luxcrypto and build libluxcrypto'
      );
    }
    const [pk, sk] = _luxcrypto.mlkem768.keypair();
    return new PQKeyExchange(pk, sk);
  }

  static fromPublicKey(publicKey: Uint8Array): PQKeyExchange {
    if (publicKey.length !== MLKEM_PUBLIC_KEY_SIZE) {
      throw new CryptoError(
        `Invalid ML-KEM public key size: expected ${MLKEM_PUBLIC_KEY_SIZE}, got ${publicKey.length}`
      );
    }
    return new PQKeyExchange(publicKey, null);
  }

  get publicKey(): Uint8Array {
    return this._publicKey;
  }

  async encapsulate(recipientPk: Uint8Array): Promise<[Uint8Array, Uint8Array]> {
    if (!_pqAvailable || !_luxcrypto) {
      throw new CryptoError('PQ crypto not available');
    }
    if (recipientPk.length !== MLKEM_PUBLIC_KEY_SIZE) {
      throw new CryptoError(
        `Invalid recipient public key size: expected ${MLKEM_PUBLIC_KEY_SIZE}, got ${recipientPk.length}`
      );
    }
    const [ct, ss] = _luxcrypto.mlkem768.encapsulate(recipientPk);
    return [ct, ss];
  }

  async decapsulate(ciphertext: Uint8Array): Promise<Uint8Array> {
    if (!_pqAvailable || !_luxcrypto) {
      throw new CryptoError('PQ crypto not available');
    }
    if (this._secretKey === null) {
      throw new CryptoError('No secret key available for decapsulation');
    }
    if (ciphertext.length !== MLKEM_CIPHERTEXT_SIZE) {
      throw new CryptoError(
        `Invalid ML-KEM ciphertext size: expected ${MLKEM_CIPHERTEXT_SIZE}, got ${ciphertext.length}`
      );
    }
    return _luxcrypto.mlkem768.decapsulate(this._secretKey, ciphertext);
  }
}

// ── ML-DSA-65 (FIPS 204) ────────────────────────────────────────────

/**
 * ML-DSA-65 Digital Signature Algorithm.
 *
 * Implements NIST FIPS 204 ML-DSA-65 via libluxcrypto (cloudflare/circl).
 * Security level: NIST Level 3 (~AES-192 equivalent).
 */
export class PQSignature {
  private readonly _publicKey: Uint8Array;
  private readonly _secretKey: Uint8Array | null;

  private constructor(publicKey: Uint8Array, secretKey: Uint8Array | null) {
    this._publicKey = publicKey;
    this._secretKey = secretKey;
  }

  static async generate(): Promise<PQSignature> {
    if (!_pqAvailable || !_luxcrypto) {
      throw new CryptoError(
        'PQ crypto not available — install luxcrypto and build libluxcrypto'
      );
    }
    const [pk, sk] = _luxcrypto.mldsa65.keypair();
    return new PQSignature(pk, sk);
  }

  static fromPublicKey(publicKey: Uint8Array): PQSignature {
    if (publicKey.length !== MLDSA_PUBLIC_KEY_SIZE) {
      throw new CryptoError(
        `Invalid ML-DSA public key size: expected ${MLDSA_PUBLIC_KEY_SIZE}, got ${publicKey.length}`
      );
    }
    return new PQSignature(publicKey, null);
  }

  get publicKey(): Uint8Array {
    return this._publicKey;
  }

  async sign(message: Uint8Array): Promise<Uint8Array> {
    if (!_pqAvailable || !_luxcrypto) {
      throw new CryptoError('PQ crypto not available');
    }
    if (this._secretKey === null) {
      throw new CryptoError('No secret key available for signing');
    }
    return _luxcrypto.mldsa65.sign(this._secretKey, message);
  }

  async verify(message: Uint8Array, signature: Uint8Array): Promise<boolean> {
    if (!_pqAvailable || !_luxcrypto) {
      throw new CryptoError('PQ crypto not available');
    }
    if (signature.length !== MLDSA_SIGNATURE_SIZE) {
      throw new CryptoError(
        `Invalid ML-DSA signature size: expected ${MLDSA_SIGNATURE_SIZE}, got ${signature.length}`
      );
    }
    return _luxcrypto.mldsa65.verify(this._publicKey, message, signature);
  }
}

// ── Hybrid Handshake ─────────────────────────────────────────────────

export type HandshakeRole = 'initiator' | 'responder';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type WebCryptoKey = any;

/**
 * Hybrid X25519 + ML-KEM-768 Handshake.
 *
 * Combines classical elliptic curve Diffie-Hellman (X25519) with post-quantum
 * ML-KEM-768 for defense-in-depth. X25519 via Web Crypto, ML-KEM via luxcrypto.
 *
 * Final shared secret derived using HKDF-SHA256.
 */
export class HybridHandshake {
  private _x25519Private: WebCryptoKey | null;
  private readonly _x25519Public: Uint8Array;
  private readonly _mlkem: PQKeyExchange;
  private readonly _role: HandshakeRole;

  private constructor(
    x25519Private: WebCryptoKey | null,
    x25519Public: Uint8Array,
    mlkem: PQKeyExchange,
    role: HandshakeRole
  ) {
    this._x25519Private = x25519Private;
    this._x25519Public = x25519Public;
    this._mlkem = mlkem;
    this._role = role;
  }

  static async initiate(): Promise<HybridHandshake> {
    if (!isWebCryptoAvailable()) {
      throw new CryptoError('Web Crypto API not available');
    }
    if (!isPQAvailable()) {
      throw new CryptoError('PQ crypto not available');
    }

    const x25519KeyPair = (await crypto.subtle.generateKey(
      { name: 'X25519' },
      true,
      ['deriveBits']
    )) as { publicKey: WebCryptoKey; privateKey: WebCryptoKey };

    const x25519PublicRaw = await crypto.subtle.exportKey(
      'raw',
      x25519KeyPair.publicKey
    );

    const mlkem = await PQKeyExchange.generate();

    return new HybridHandshake(
      x25519KeyPair.privateKey,
      new Uint8Array(x25519PublicRaw),
      mlkem,
      'initiator'
    );
  }

  get publicData(): HybridInitiatorData {
    return {
      x25519PublicKey: this._x25519Public,
      mlkemPublicKey: this._mlkem.publicKey,
    };
  }

  static async respond(
    initiatorData: HybridInitiatorData
  ): Promise<[HybridHandshake, HybridResponderData]> {
    if (!isWebCryptoAvailable()) {
      throw new CryptoError('Web Crypto API not available');
    }
    if (!isPQAvailable()) {
      throw new CryptoError('PQ crypto not available');
    }

    if (initiatorData.x25519PublicKey.length !== X25519_PUBLIC_KEY_SIZE) {
      throw new CryptoError(
        `Invalid X25519 public key size: expected ${X25519_PUBLIC_KEY_SIZE}, got ${initiatorData.x25519PublicKey.length}`
      );
    }
    if (initiatorData.mlkemPublicKey.length !== MLKEM_PUBLIC_KEY_SIZE) {
      throw new CryptoError(
        `Invalid ML-KEM public key size: expected ${MLKEM_PUBLIC_KEY_SIZE}, got ${initiatorData.mlkemPublicKey.length}`
      );
    }

    const x25519KeyPair = (await crypto.subtle.generateKey(
      { name: 'X25519' },
      true,
      ['deriveBits']
    )) as { publicKey: WebCryptoKey; privateKey: WebCryptoKey };

    const x25519PublicRaw = await crypto.subtle.exportKey(
      'raw',
      x25519KeyPair.publicKey
    );

    const mlkem = await PQKeyExchange.generate();
    const [mlkemCiphertext] = await mlkem.encapsulate(initiatorData.mlkemPublicKey);

    const response: HybridResponderData = {
      x25519PublicKey: new Uint8Array(x25519PublicRaw),
      mlkemCiphertext,
    };

    const handshake = new HybridHandshake(
      x25519KeyPair.privateKey,
      new Uint8Array(x25519PublicRaw),
      mlkem,
      'responder'
    );

    return [handshake, response];
  }

  async finalize(responderData: HybridResponderData): Promise<Uint8Array> {
    if (this._role !== 'initiator') {
      throw new CryptoError('finalize() can only be called by initiator');
    }
    if (this._x25519Private === null) {
      throw new CryptoError('X25519 private key not available');
    }

    const peerX25519Public = await crypto.subtle.importKey(
      'raw',
      responderData.x25519PublicKey,
      { name: 'X25519' },
      false,
      []
    );

    const x25519Shared = await crypto.subtle.deriveBits(
      { name: 'X25519', public: peerX25519Public },
      this._x25519Private,
      256
    );

    const mlkemShared = await this._mlkem.decapsulate(responderData.mlkemCiphertext);

    this._x25519Private = null;

    return this.deriveHybridSecret(new Uint8Array(x25519Shared), mlkemShared);
  }

  async complete(
    initiatorData: HybridInitiatorData,
    mlkemShared?: Uint8Array
  ): Promise<Uint8Array> {
    if (this._role !== 'responder') {
      throw new CryptoError('complete() can only be called by responder');
    }
    if (this._x25519Private === null) {
      throw new CryptoError('X25519 private key not available');
    }

    const peerX25519Public = await crypto.subtle.importKey(
      'raw',
      initiatorData.x25519PublicKey,
      { name: 'X25519' },
      false,
      []
    );

    const x25519Shared = await crypto.subtle.deriveBits(
      { name: 'X25519', public: peerX25519Public },
      this._x25519Private,
      256
    );

    let finalMlkemShared = mlkemShared;
    if (!finalMlkemShared) {
      [, finalMlkemShared] = await this._mlkem.encapsulate(initiatorData.mlkemPublicKey);
    }

    this._x25519Private = null;

    return this.deriveHybridSecret(new Uint8Array(x25519Shared), finalMlkemShared);
  }

  private async deriveHybridSecret(
    x25519Shared: Uint8Array,
    mlkemShared: Uint8Array
  ): Promise<Uint8Array> {
    const ikm = new Uint8Array(x25519Shared.length + mlkemShared.length);
    ikm.set(x25519Shared);
    ikm.set(mlkemShared, x25519Shared.length);

    const ikmKey = await crypto.subtle.importKey(
      'raw',
      ikm,
      { name: 'HKDF' },
      false,
      ['deriveBits']
    );

    const salt = new TextEncoder().encode('ZAP-HYBRID-HANDSHAKE-v1');
    const info = new TextEncoder().encode('shared-secret');

    const derived = await crypto.subtle.deriveBits(
      {
        name: 'HKDF',
        hash: 'SHA-256',
        salt,
        info,
      },
      ikmKey,
      HYBRID_SHARED_SECRET_SIZE * 8
    );

    return new Uint8Array(derived);
  }
}

/**
 * Perform a complete hybrid handshake between two parties.
 */
export async function hybridHandshake(): Promise<[Uint8Array, Uint8Array]> {
  const initiator = await HybridHandshake.initiate();
  const initData = initiator.publicData;

  const [responder, respData] = await HybridHandshake.respond(initData);

  const mlkem = await PQKeyExchange.generate();
  const [, mlkemShared] = await mlkem.encapsulate(initData.mlkemPublicKey);

  const initiatorSecret = await initiator.finalize(respData);
  const responderSecret = await responder.complete(initData, mlkemShared);

  return [initiatorSecret, responderSecret];
}

File: src/crypto.rs
//! Post-Quantum Cryptography Module for ZAP
//!
//! One library, one implementation: pure-Rust pqcrypto crates.
//! ML-KEM-768 (FIPS 203) for key exchange and Dilithium3 for signatures,
//! gated behind the `pq` feature. X25519 always uses x25519-dalek.
//!
//! # Example
//!
//! ```rust,ignore
//! use zap::crypto::{PQKeyExchange, PQSignature, HybridHandshake};
//!
//! // Key exchange
//! let alice = PQKeyExchange::generate()?;
//! let (ciphertext, shared_alice) = alice.encapsulate(&bob_pk)?;
//! let shared_bob = bob.decapsulate(&ciphertext)?;
//! assert_eq!(shared_alice, shared_bob);
//!
//! // Signatures
//! let signer = PQSignature::generate()?;
//! let sig = signer.sign(b"message")?;
//! signer.verify(b"message", &sig)?;
//!
//! // Hybrid handshake
//! let initiator = HybridHandshake::initiate()?;
//! let (responder, response) = HybridHandshake::respond(&initiator.public_data())?;
//! let shared = initiator.finalize(&response)?;
//! ```

use crate::error::{Error, Result};

/// ML-KEM-768 public key size in bytes
pub const MLKEM_PUBLIC_KEY_SIZE: usize = 1184;
/// ML-KEM-768 ciphertext size in bytes
pub const MLKEM_CIPHERTEXT_SIZE: usize = 1088;
/// ML-KEM-768 shared secret size in bytes
pub const MLKEM_SHARED_SECRET_SIZE: usize = 32;
/// ML-DSA-65 public key size in bytes
pub const MLDSA_PUBLIC_KEY_SIZE: usize = 1952;
/// ML-DSA-65 signature size in bytes
pub const MLDSA_SIGNATURE_SIZE: usize = 3309;
/// ML-DSA-65 secret key size in bytes
pub const MLDSA_SECRET_KEY_SIZE: usize = 4032;
/// X25519 public key size in bytes
pub const X25519_PUBLIC_KEY_SIZE: usize = 32;
/// Hybrid shared secret size after HKDF
pub const HYBRID_SHARED_SECRET_SIZE: usize = 32;

// ── Backend detection ────────────────────────────────────────────────

/// Which PQ backend is active.
///
/// - `"pq"` — Rust pqcrypto crates (ML-KEM-768 + Dilithium3)
/// - `"unavailable"` — no PQ feature enabled
pub const PQ_BACKEND: &str = if cfg!(feature = "pq") {
    "pq"
} else {
    "unavailable"
};

// ── pqcrypto backend ─────────────────────────────────────────────────

#[cfg(feature = "pq")]
mod pq_impl {
    use super::*;
    use hkdf::Hkdf;
    use pqcrypto_dilithium::dilithium3;
    use pqcrypto_mlkem::mlkem768;
    use pqcrypto_traits::kem::{Ciphertext, PublicKey as KemPublicKey, SharedSecret};
    use pqcrypto_traits::sign::{
        DetachedSignature as DetachedSignatureTrait, PublicKey as SignPublicKey,
        SecretKey as SignSecretKey,
    };
    use rand::rngs::OsRng;
    use sha2::Sha256;
    use x25519_dalek::{EphemeralSecret, PublicKey as X25519PublicKey};
    use zeroize::Zeroize;

    pub struct PQKeyExchange {
        public_key: mlkem768::PublicKey,
        secret_key: mlkem768::SecretKey,
    }

    impl PQKeyExchange {
        pub fn generate() -> Result<Self> {
            let (pk, sk) = mlkem768::keypair();
            Ok(Self {
                public_key: pk,
                secret_key: sk,
            })
        }

        pub fn public_key_bytes(&self) -> Vec<u8> {
            self.public_key.as_bytes().to_vec()
        }

        pub fn from_public_key(bytes: &[u8]) -> Result<Self> {
            if bytes.len() != MLKEM_PUBLIC_KEY_SIZE {
                return Err(Error::Crypto(format!(
                    "invalid ML-KEM public key size: expected {}, got {}",
                    MLKEM_PUBLIC_KEY_SIZE,
                    bytes.len()
                )));
            }
            let pk = mlkem768::PublicKey::from_bytes(bytes)
                .map_err(|e| Error::Crypto(format!("invalid ML-KEM public key: {e:?}")))?;
            let (_, dummy_sk) = mlkem768::keypair();
            Ok(Self {
                public_key: pk,
                secret_key: dummy_sk,
            })
        }

        pub fn encapsulate(&self, recipient_pk: &[u8]) -> Result<(Vec<u8>, [u8; 32])> {
            let pk = mlkem768::PublicKey::from_bytes(recipient_pk)
                .map_err(|e| Error::Crypto(format!("invalid recipient public key: {e:?}")))?;
            let (ss, ct) = mlkem768::encapsulate(&pk);
            let mut shared = [0u8; 32];
            shared.copy_from_slice(ss.as_bytes());
            Ok((ct.as_bytes().to_vec(), shared))
        }

        pub fn decapsulate(&self, ciphertext: &[u8]) -> Result<[u8; 32]> {
            if ciphertext.len() != MLKEM_CIPHERTEXT_SIZE {
                return Err(Error::Crypto(format!(
                    "invalid ML-KEM ciphertext size: expected {}, got {}",
                    MLKEM_CIPHERTEXT_SIZE,
                    ciphertext.len()
                )));
            }
            let ct = mlkem768::Ciphertext::from_bytes(ciphertext)
                .map_err(|e| Error::Crypto(format!("invalid ciphertext: {e:?}")))?;
            let ss = mlkem768::decapsulate(&ct, &self.secret_key);
            let mut shared = [0u8; 32];
            shared.copy_from_slice(ss.as_bytes());
            Ok(shared)
        }
    }

    pub struct PQSignature {
        public_key: dilithium3::PublicKey,
        secret_key: Option<dilithium3::SecretKey>,
    }

    impl std::fmt::Debug for PQSignature {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            f.debug_struct("PQSignature")
                .field("public_key", &"<public_key>")
                .field(
                    "secret_key",
                    &self.secret_key.as_ref().map(|_| "<secret_key>"),
                )
                .finish()
        }
    }

    impl Clone for PQSignature {
        fn clone(&self) -> Self {
            let pk_bytes = self.public_key.as_bytes().to_vec();
            let public_key = dilithium3::PublicKey::from_bytes(&pk_bytes).unwrap();
            let secret_key = self.secret_key.as_ref().map(|sk| {
                let sk_bytes = sk.as_bytes().to_vec();
                dilithium3::SecretKey::from_bytes(&sk_bytes).unwrap()
            });
            Self {
                public_key,
                secret_key,
            }
        }
    }

    impl PQSignature {
        pub fn generate() -> Result<Self> {
            let (pk, sk) = dilithium3::keypair();
            Ok(Self {
                public_key: pk,
                secret_key: Some(sk),
            })
        }

        pub fn public_key_bytes(&self) -> Vec<u8> {
            self.public_key.as_bytes().to_vec()
        }

        pub fn from_public_key(bytes: &[u8]) -> Result<Self> {
            if bytes.len() != MLDSA_PUBLIC_KEY_SIZE {
                return Err(Error::Crypto(format!(
                    "invalid ML-DSA public key size: expected {}, got {}",
                    MLDSA_PUBLIC_KEY_SIZE,
                    bytes.len()
                )));
            }
            let pk = dilithium3::PublicKey::from_bytes(bytes)
                .map_err(|e| Error::Crypto(format!("invalid ML-DSA public key: {e:?}")))?;
            Ok(Self {
                public_key: pk,
                secret_key: None,
            })
        }

        pub fn sign(&self, message: &[u8]) -> Result<Vec<u8>> {
            let sk = self
                .secret_key
                .as_ref()
                .ok_or_else(|| Error::Crypto("no secret key available for signing".into()))?;
            let sig = dilithium3::detached_sign(message, sk);
            Ok(sig.as_bytes().to_vec())
        }

        pub fn verify(&self, message: &[u8], signature: &[u8]) -> Result<()> {
            if signature.len() != MLDSA_SIGNATURE_SIZE {
                return Err(Error::Crypto(format!(
                    "invalid ML-DSA signature size: expected {}, got {}",
                    MLDSA_SIGNATURE_SIZE,
                    signature.len()
                )));
            }
            let sig = dilithium3::DetachedSignature::from_bytes(signature)
                .map_err(|e| Error::Crypto(format!("invalid signature format: {e:?}")))?;
            dilithium3::verify_detached_signature(&sig, message, &self.public_key)
                .map_err(|_| Error::Crypto("signature verification failed".into()))
        }
    }

    #[derive(Debug, Clone)]
    pub struct HybridInitiatorData {
        pub x25519_public_key: [u8; 32],
        pub mlkem_public_key: Vec<u8>,
    }

    #[derive(Debug, Clone)]
    pub struct HybridResponderData {
        pub x25519_public_key: [u8; 32],
        pub mlkem_ciphertext: Vec<u8>,
    }

    #[derive(Clone)]
    pub struct HybridSharedSecret {
        secret: [u8; HYBRID_SHARED_SECRET_SIZE],
    }

    impl HybridSharedSecret {
        pub fn as_bytes(&self) -> &[u8; HYBRID_SHARED_SECRET_SIZE] {
            &self.secret
        }
        pub fn into_bytes(self) -> [u8; HYBRID_SHARED_SECRET_SIZE] {
            self.secret
        }
    }

    impl Drop for HybridSharedSecret {
        fn drop(&mut self) {
            self.secret.zeroize();
        }
    }

    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    enum HandshakeRole {
        Initiator,
        Responder,
    }

    pub struct HybridHandshake {
        x25519_secret: Option<EphemeralSecret>,
        x25519_public: X25519PublicKey,
        mlkem: PQKeyExchange,
        role: HandshakeRole,
    }

    impl HybridHandshake {
        pub fn initiate() -> Result<Self> {
            let x25519_secret = EphemeralSecret::random_from_rng(OsRng);
            let x25519_public = X25519PublicKey::from(&x25519_secret);
            let mlkem = PQKeyExchange::generate()?;
            Ok(Self {
                x25519_secret: Some(x25519_secret),
                x25519_public,
                mlkem,
                role: HandshakeRole::Initiator,
            })
        }

        pub fn public_data(&self) -> HybridInitiatorData {
            HybridInitiatorData {
                x25519_public_key: self.x25519_public.to_bytes(),
                mlkem_public_key: self.mlkem.public_key_bytes(),
            }
        }

        pub fn respond(
            initiator_data: &HybridInitiatorData,
        ) -> Result<(Self, HybridResponderData)> {
            if initiator_data.mlkem_public_key.len() != MLKEM_PUBLIC_KEY_SIZE {
                return Err(Error::Crypto(format!(
                    "invalid initiator ML-KEM public key size: expected {}, got {}",
                    MLKEM_PUBLIC_KEY_SIZE,
                    initiator_data.mlkem_public_key.len()
                )));
            }
            let x25519_secret = EphemeralSecret::random_from_rng(OsRng);
            let x25519_public = X25519PublicKey::from(&x25519_secret);
            let mlkem = PQKeyExchange::generate()?;
            let (mlkem_ciphertext, _) = mlkem.encapsulate(&initiator_data.mlkem_public_key)?;
            let response = HybridResponderData {
                x25519_public_key: x25519_public.to_bytes(),
                mlkem_ciphertext,
            };
            let handshake = Self {
                x25519_secret: Some(x25519_secret),
                x25519_public,
                mlkem,
                role: HandshakeRole::Responder,
            };
            Ok((handshake, response))
        }

        pub fn finalize(
            mut self,
            responder_data: &HybridResponderData,
        ) -> Result<HybridSharedSecret> {
            if self.role != HandshakeRole::Initiator {
                return Err(Error::Crypto(
                    "finalize() can only be called by initiator".into(),
                ));
            }
            let x25519_secret = self
                .x25519_secret
                .take()
                .ok_or_else(|| Error::Crypto("X25519 secret already consumed".into()))?;
            let peer = X25519PublicKey::from(responder_data.x25519_public_key);
            let x25519_shared = x25519_secret.diffie_hellman(&peer);
            let mlkem_shared = self.mlkem.decapsulate(&responder_data.mlkem_ciphertext)?;
            Self::derive_hybrid_secret(x25519_shared.as_bytes(), &mlkem_shared)
        }

        pub fn complete(
            mut self,
            initiator_data: &HybridInitiatorData,
            mlkem_shared: &[u8; 32],
        ) -> Result<HybridSharedSecret> {
            if self.role != HandshakeRole::Responder {
                return Err(Error::Crypto(
                    "complete() can only be called by responder".into(),
                ));
            }
            let x25519_secret = self
                .x25519_secret
                .take()
                .ok_or_else(|| Error::Crypto("X25519 secret already consumed".into()))?;
            let peer = X25519PublicKey::from(initiator_data.x25519_public_key);
            let x25519_shared = x25519_secret.diffie_hellman(&peer);
            Self::derive_hybrid_secret(x25519_shared.as_bytes(), mlkem_shared)
        }

        fn derive_hybrid_secret(
            x25519_shared: &[u8],
            mlkem_shared: &[u8; 32],
        ) -> Result<HybridSharedSecret> {
            let mut ikm = Vec::with_capacity(x25519_shared.len() + mlkem_shared.len());
            ikm.extend_from_slice(x25519_shared);
            ikm.extend_from_slice(mlkem_shared);
            let hkdf = Hkdf::<Sha256>::new(Some(b"ZAP-HYBRID-HANDSHAKE-v1"), &ikm);
            let mut secret = [0u8; HYBRID_SHARED_SECRET_SIZE];
            hkdf.expand(b"shared-secret", &mut secret)
                .map_err(|_| Error::Crypto("HKDF expansion failed".into()))?;
            ikm.zeroize();
            Ok(HybridSharedSecret { secret })
        }
    }

    pub fn hybrid_handshake() -> Result<(
        [u8; HYBRID_SHARED_SECRET_SIZE],
        [u8; HYBRID_SHARED_SECRET_SIZE],
    )> {
        let initiator = HybridHandshake::initiate()?;
        let init_data = initiator.public_data();
        let (responder, resp_data) = HybridHandshake::respond(&init_data)?;
        let mlkem_for_responder = PQKeyExchange::generate()?;
        let (_, mlkem_shared_responder) =
            mlkem_for_responder.encapsulate(&init_data.mlkem_public_key)?;
        let initiator_secret = initiator.finalize(&resp_data)?;
        let responder_secret = responder.complete(&init_data, &mlkem_shared_responder)?;
        Ok((initiator_secret.into_bytes(), responder_secret.into_bytes()))
    }

    #[cfg(test)]
    mod tests {
        use super::*;

        #[test]
        fn test_mlkem_key_exchange() {
            let alice = PQKeyExchange::generate().unwrap();
            let bob = PQKeyExchange::generate().unwrap();
            let (ciphertext, alice_shared) = alice.encapsulate(&bob.public_key_bytes()).unwrap();
            let bob_shared = bob.decapsulate(&ciphertext).unwrap();
            assert_eq!(alice_shared, bob_shared);
        }

        #[test]
        fn test_mldsa_signature() {
            let signer = PQSignature::generate().unwrap();
            let message = b"The quick brown fox jumps over the lazy dog";
            let signature = signer.sign(message).unwrap();
            signer.verify(message, &signature).unwrap();
            let verifier = PQSignature::from_public_key(&signer.public_key_bytes()).unwrap();
            verifier.verify(message, &signature).unwrap();
        }
    }
}

// ── Re-exports ───────────────────────────────────────────────────────

#[cfg(feature = "pq")]
pub use pq_impl::{
    hybrid_handshake, HybridHandshake, HybridInitiatorData, HybridResponderData,
    HybridSharedSecret, PQKeyExchange, PQSignature,
};

// Feature-gated fallbacks when the `pq` feature is not enabled
#[cfg(not(feature = "pq"))]
pub struct PQKeyExchange;

#[cfg(not(feature = "pq"))]
impl PQKeyExchange {
    pub fn generate() -> Result<Self> {
        Err(Error::Crypto("PQ crypto requires 'pq' feature".into()))
    }
}

#[cfg(not(feature = "pq"))]
pub struct PQSignature;

#[cfg(not(feature = "pq"))]
impl PQSignature {
    pub fn generate() -> Result<Self> {
        Err(Error::Crypto("PQ crypto requires 'pq' feature".into()))
    }
}

#[cfg(not(feature = "pq"))]
pub struct HybridHandshake;

#[cfg(not(feature = "pq"))]
impl HybridHandshake {
    pub fn initiate() -> Result<Self> {
        Err(Error::Crypto("PQ crypto requires 'pq' feature".into()))
    }
}
